### PR TITLE
Add onboarding to master stack

### PR DESCRIPTION
**Pull request by austinbyers**
## Background

There was quite a bit of `mage` code surrounding the onboard stack, both before and after deploying it. These have all been moved to custom resources

Will likely conflict with #949 , I'll wait for that to merge first

## Changes

- Add 3 new custom resources for the onboarding stack
    - `Custom::SelfLogConfig` - S3/GuardDuty settings which aren't supported in CloudFormation
    - `Custom::SelfRegistration` - Hits the source-api to register Panther cloud security / log analysis as sources
    - `Custom::StackSet` - Deploy/Teardown a stack set with a single instance in the current region. This is used for real-time cloud security events, but is generic enough to have other potential applications in the future

## Testing

- Deploy/teardown with mage
- Deploy/teardown with master template
- Toggled OnboardSelf, EnableGuardDuty, etc

### CODE REVIEW

File: internal/core/custom_resources/resources/self_registration.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			registerCloudSec = false
+		}
+		if aws.StringValue(integration.AWSAccountID) == props.AccountID &&
+			*integration.IntegrationType == models.IntegrationTypeAWS3 &&
+			*integration.IntegrationLabel == genLogProcessingLabel() {
+
+			zap.L().Info("account already registered for log processing",
+				zap.String("accountID", props.AccountID))
+			registerLogProcessing = false
+		}
```
Comment (kostaspap) [inline internal/core/custom_resources/resources/self_registration.go:95]: This logic could cause issues in the following scenario: 
1. Someone deploys panther and enables the ALB and VPC FL. 
1. They decided to turn S3 Server Access on. They re-deploy Panther. 

This logic will not update the the log types of the integration, right?

File: deployments/onboard.yml [hunk 2]
```diff
@@ -109,7 +113,7 @@ Resources:
   LogProcessingRole:
     Type: AWS::IAM::Role
     Properties:
-      RoleName: !Sub PantherLogProcessingRole-${LogProcessingRoleLabel}
+      RoleName: !Sub PantherLogProcessingRole-panther-account-${AWS::Region}
```
Comment (kostaspap) [inline deployments/onboard.yml:116]: If you could take this role out of this template and create it in a new stack using `panther/deployments/auxiliary/cloudformation/panther-log-analysis-iam.yml` with appropriate parameters, I think it would solve this long standing issue as well: https://github.com/panther-labs/panther/issues/640

Not sure if it's easy to do as part of this PR.

File: internal/core/custom_resources/resources/self_log_config.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	case cfn.RequestCreate, cfn.RequestUpdate:
+		var props SelfLogConfigProperties
+		if err := parseProperties(event.ResourceProperties, &props); err != nil {
+			return "", nil, err
+		}
+		return "custom:self-log-config:singleton", nil, configureLogProcessingUsingAPIs(props)
+
+	default:
+		// ignore deletes
+		return event.PhysicalResourceID, nil, nil
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/core/custom_resources/resources/self_log_config.go:51]: Good point. I think today if you disable the self onboarding the registrations are not removed. The CF took care of the permissions and CF constructed entities, so as a practical matter it may not matter.

We need to preserve at least the same clean up as what was happening in CF. For example, un-registering GuardDuty.

File: internal/core/custom_resources/resources/self_registration.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	switch event.RequestType {
+	case cfn.RequestCreate, cfn.RequestUpdate:
+		var props SelfRegistrationProperties
+		if err := parseProperties(event.ResourceProperties, &props); err != nil {
+			return "", nil, err
+		}
+		return "custom:self-registration:singleton", nil, registerPantherAccount(props)
+
+	default:
+		// ignore deletes
```
Comment (rleighton) [inline internal/core/custom_resources/resources/self_registration.go:60]: Per Kostas' note above, don't we need to delete the self integrations when they disable the self monitoring?

File: internal/core/custom_resources/resources/self_registration.go (hunk 1 above)
Comment (rleighton) [reply internal/core/custom_resources/resources/self_registration.go:95]: My PR handles that, I fixed that. Austin can copy that part into this.

File: internal/core/custom_resources/resources/stack_set.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		}
+
+		return event.PhysicalResourceID, nil, deleteStackSet(split[3], split[4])
+
+	default:
+		return "", nil, fmt.Errorf("unknown request type %s", event.RequestType)
+	}
+}
+
+// Returns once parent stack set has started creation (does not wait for it to finish).
```
Comment (rleighton) [inline internal/core/custom_resources/resources/stack_set.go:70]: Is that what mage does today? How do we know about errors?

File: deployments/onboard.yml (hunk 2 above)
Comment (austinbyers) [reply deployments/onboard.yml:116]: Interesting. I don't fully understand what's needed for that issue, what would I use for the `RoleSuffix`? Can you show exactly how I should be instantiating that stack?

```yaml
# auxiliary/cloudformation/panther-log-anslysis-iam.yml
Mappings:
  # DO NOT EDIT PantherParameters section. Panther application relies on the exact format (including comments)
  # in order to replace the default values with an appropriate ones.
  PantherParameters:
    MasterAccountId:
      Value: '' # MasterAccountId
    RoleSuffix:
      Value: '' # RoleSuffix

Parameters:
  # Required parameters
  MasterAccountId:
    Type: String
    Description: DO NOT EDIT MANUALLY! Parameter is already populated with the appropriate value.
    Default: ''
  RoleSuffix:
    Type: String
    Description: DO NOT EDIT MANUALLY! Parameter is already populated with the appropriate value.
    Default: ''

Conditions:
  # Condition to define if the template is generated by panther backend
  IsGenerated: !Not [!Equals ['', !FindInMap [PantherParameters, MasterAccountId, Value]]]

Resources:
  LogProcessingRole:
    Type: AWS::IAM::Role
    Properties:
      RoleName: !If
        - IsGenerated
        - !Sub
          - 'PantherLogProcessingRole-${Suffix}'
          - Suffix: !FindInMap [PantherParameters, RoleSuffix, Value]
        - !Sub 'PantherLogProcessingRole-${RoleSuffix}'
```

If the IAM role will have the same name, then we'll have a conflict - CloudFormation will destroy this IAM role and create a new one in the nested stack with the same name at the same time, which will probably fail. We'd have to add a migration to `mage deploy` to delete the existing role before deploying the stacks, which will also temporarily break log processing I assume?

If the role name will be different, there's no conflicts and I'm happy to make that change now. If the role name is the same and we need a migration, I'll just add a comment and we'll do that in a different PR

Comment (austinbyers) [reply deployments/onboard.yml:116]: The issue you cited (#640) mentions unpredictable/inconsistent stack names, how would this help solve that?

Comment (kostaspap) [reply deployments/onboard.yml:116]: You can pass in the parameters to the CFN stack. 
- MasterAccountId: The account id where panther is deployed
- RoleSuffix : The `IntegrationLabel`, lowercased and with `-` instead of ` `. In this case `panther-account`.

You are right, to do the migration we should delete first the old role then create the new one. Or change the label (and `RoleSuffix` as a consequence) to something else e.g. `panther-account-onboard` and then delete the old integration (with label `panther-account`). If a role doesn't exist for a bit is fine - there will be no data loss, the log processor retries to pull the logs for a while (a couple of hours). Feel free to skip these changes and add a TODO or something if you think it's more appropriate. 

This template has a very very ugly fix to allow us to create/update templates both from UI and templates that are downloaded from the UI then deployed manually :(

File: internal/core/custom_resources/resources/self_log_config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (austinbyers) [reply internal/core/custom_resources/resources/self_log_config.go:51]: > Is the "delete" going to be invoked when someone switches from from OnboardSelf: true to OnboardSelf: false?

When using the master stack, yes. The nested `onboard` stack would be automatically deleted, which would then delete all custom resources as well. 

> I think today if you disable the self onboarding the registrations are not removed.

Corect - when deploying from mage, we just say "if OnboardSelf, deploy(onboardStack)". The stack in that case would NOT be destroyed if OnboardSelf is toggled off, but now that's a fairly easy change I can make.

This is a great example of how using a custom resources makes our cleanup automatically better than it was before - today, `mage` only deletes onboarding resources during a teardown.

However, these steps do not yet have special cleanup:

* `Custom::SelfRegistration` - we could deregister Panther from the source-api
* `Custom::SelfLogConfig`
    * GuardDuty publishing destination - when the detector is destroyed, this presumably won't matter anyway
    * S3 bucket notification - we could remove the notification

I may actually split the `SelfLogConfig` into two separate resources, then, to better handle cleanup.

-----
In summary:

* Using a custom resource means our cleanup is already better than it used to be
* I'll fix `mage deploy` to teardown self-onboarding if `OnboardSelf=false`
* I'll add cleanup for the other custom resources (which we did not have before in `mage`)

File: internal/core/custom_resources/resources/self_registration.go (hunk 4 above)
Comment (austinbyers) [reply internal/core/custom_resources/resources/self_registration.go:60]: Sure, good idea. We don't have this in mage today, and I just ported what we had before, but now that's a relatively easy change I can make

File: deployments/onboard.yml (hunk 2 above)
Comment (kostaspap) [reply deployments/onboard.yml:116]: > The issue you cited (#640) mentions unpredictable/inconsistent stack names, how would this help solve that?

With the change above, the stack names will be with what we do normally during S3 onboarding!
Normally, the IAM role is deployed to its own stack and has a specific name. In case of the "onboard" stack, the IAM role was(is) nested

File: internal/core/custom_resources/resources/stack_set.go (hunk 5 above)
Comment (austinbyers) [reply internal/core/custom_resources/resources/stack_set.go:70]: Hmm good point. `mage` does wait for the stack set to finish creating and destroying.

I wasn't sure if it was a good idea to have the custom resource waiting while the stack sets finished, because of the timeout and because you maxed out the memory. If there was an error *starting* the stack set, the deploy would fail. But you're right that if there's an error finishing it, the deploy process would never know.

I'll go ahead and put the wait back in, but then I will also reduce the memory to, say, 1024 so we're not paying the maximum price when waiting for 10-120 seconds for every deploy + teardown

File: deployments/onboard.yml (hunk 2 above)
Comment (austinbyers) [reply deployments/onboard.yml:116]: Ok, if deleting the role and re-creating it is fine, that's fairly easy and I can probably add that.

> With the change above, the stack names will be with what we do normally during S3 onboarding!

With the introduction of the master stack, we don't have any control over the stack names. Someone can deploy that master template with any stack name they choose

File: internal/core/custom_resources/resources/stack_set.go (hunk 5 above)
Comment (austinbyers) [reply internal/core/custom_resources/resources/stack_set.go:70]: Ah, I was wrong - `mage deploy` today does not wait for the stack set instances to finish (but it does wait for teardown). I'll add it nonetheless

File: deployments/onboard.yml (hunk 2 above)
Comment (austinbyers) [reply deployments/onboard.yml:116]: More precisely, to answer this:

> If you could take this role out of this template and create it in a new stack 

We have to give people a single master template, we can't have some IAM roles created in a different top-level stack for packaged deployments

File: internal/core/custom_resources/resources/self_registration.go (hunk 1 above)
Comment (austinbyers) [reply internal/core/custom_resources/resources/self_registration.go]: Done

File: deployments/onboard.yml (hunk 2 above)
Comment (austinbyers) [reply deployments/onboard.yml]: I have switched the log role to use the aux template. Because the logicalID is the same, I actually don't think any special migration is necessary (but I will test)

File: internal/core/custom_resources/resources/stack_set.go (hunk 5 above)
Comment (austinbyers) [reply internal/core/custom_resources/resources/stack_set.go]: On second thought, I've actually removed the stackset altogether (see updated PR description / comment)

**Comment by austinbyers** [comment]
@rleighton @kostaspap please take another look! Significant improvements since your last review:

- Better teardown/cleanup:
    - `mage deploy` now deletes the onboard stack if OnboardSelf=false
    - When `Custom::SelfRegistration` is deleted, it will remove the sources from source-api
    - `Custom::SelfLogConfig` has been split into its fundamental components: `Custom::S3BucketNotification` and `Custom::GuardDutyDestination`. This way, the custom resources are more generic, we can use the AWS Go structs as the custom resource properties, and there is a clean delete implementation for each
- Log processing role has been switched to use the aux template (instead of being defined directly in the onboarding stack)
- The real-time events stack set has been replaced with a simple nested template. Stack sets help with deploying across multiple accounts or regions, but the default self-onboarding does neither of these things - it uses a single stack instance in a single region. There are only 4 resources in the `cloudwatch-events` stack, all of which are safe to be deployed in multiple regions.
    - Trying to manage the stack set with a custom resource is possible, but way overkill for our needs. We have to manage the stack set instance and the parent and wait for both to finish - hundreds of lines of code and wasted custom-resource lambda cycles
- Moved migrations into a separate file (and added the stack set deletion as a new migration)

## Testing
- Deployed from mage to us-west-2
- Deployed from master template to us-east-1
- Upgrade us-east-2 1.3.0 deployment (to test migration)

File: deployments/onboard.yml (hunk 2 above)
Comment (austinbyers) [reply deployments/onboard.yml]: And I was wrong:

```
failed to create change set for stack panther-onboard: ValidationError: Update of resource type is not permitted. The new template modifies resource type of the following resources: [LogProcessingRole]
```

Migration added

File: tools/mage/migrations.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+	// Removed stacks
+	alarmsStack            = "panther-cw-alarms"
+	metricFilterStack      = "panther-cw-metric-filters"
+	realTimeEventsStackSet = "panther-real-time-events"
+)
+
+// Migrations which run before the main deploy.
+//
+// These can be removed a few releases after they have been added.
+func migrate(awsSession *session.Session, accountID string) {
```
Comment (kostaspap) [inline tools/mage/migrations.go:42]: Nice! I was thinking that the migrations in the future could be split in chunks. E.g. the code would be checking the already deployed release (visible as a tag in the deployed CFN stacks) and running the appropriate migration required between the deploy and the latest available. A "migration" would be basically valid only for one updating from one version to the next. 

E.g. 
We can have something like:
```
type Migration interface{
  GetVersion() string
  Migration func() error
}
```
and a different migration per version. Or just `version_1_3_0.Migrate()` etc and the "main" migration code would be invoking the above in sequence. 


I know it's too early for this. This PR already improves a lot the current process, just thinking out loud how this could look in the future.

Comment (austinbyers) [reply tools/mage/migrations.go:42]: >  A "migration" would be basically valid only for one updating from one version to the next.

The problem with this approach is that developers need the migrations also. As soon as a breaking change lands in master, all developers immediately need the migration. That's why I prefer to check for the actual infrastructure markers rather than a specific version.

This method is also more robust if, say, an upgrade fails halfway through - some stacks might be v1.3.0 and other stacks might be v1.4.0, what then? It's not clear whether to run the migration.

Checking the infra the way we do now does not take more than a second or two and can be removed after a few releases

Comment (kostaspap) [reply tools/mage/migrations.go:42]: I see. Yeah indeed for such usecases this looks more robust. 

The usecases I had in mind were:
1. Managing migration long-term. E.g. not sure when is really a good time to drop some migration code (if ever for enterprise customers)
2. Migrations that are not dependent on infra e.g. running `glue:sync` equivalent code when needed, database migrations etc

File: internal/core/custom_resources/resources/self_registration.go [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+		zap.String("accountID", props.AccountID))
+
+	// avoid alarms/errors and check first if the integrations exist
+	var listOutput []*models.SourceIntegration
+	var listInput = &models.LambdaInput{
+		ListIntegrations: &models.ListIntegrationsInput{},
+	}
+	if err := genericapi.Invoke(getLambdaClient(), "panther-source-api", listInput, &listOutput); err != nil {
+		return fmt.Errorf("error calling source-api to list integrations: %v", err)
+	}
```
Comment (nhakmiller) [inline internal/core/custom_resources/resources/self_registration.go:80]: Is this code accomplishing anything? In the `getSelfIntegrations` call below you do the exact same thing anyways.

Comment (austinbyers) [reply internal/core/custom_resources/resources/self_registration.go:80]: You're right! I moved it to the helper function and forgot to delete it here. Will do
